runtime/src/accounts_db.rs

```diff
@@ -10431,76 +10431,6 @@ pub mod tests {
         do_test_load_account_and_shrink_race(false);
     }
 
-    fn do_test_load_account_and_purge_race(with_retry: bool) {
-        let caching_enabled = true;
-        let mut db = AccountsDb::new_with_config(
-            Vec::new(),
-            &ClusterType::Development,
-            AccountSecondaryIndexes::default(),
-            caching_enabled,
-        );
-        db.load_delay = RACY_SLEEP_MS;
-        let db = Arc::new(db);
-        let pubkey =
-            Arc::new(Pubkey::from_str("CiDwVBFgWV9E5MvXWoLgnEgn2hK7rJikbvfWavzAQz3").unwrap());
-        let exit = Arc::new(AtomicBool::new(false));
-        let slot = 1;
-
-        // Store an account
-        let lamports = 42;
-        let mut account = AccountSharedData::new(1, 0, AccountSharedData::default().owner());
-        account.set_lamports(lamports);
-        db.store_uncached(slot, &[(&pubkey, &account)]);
-
-        let t_purge_slot = {
-            let db = db.clone();
-            let exit = exit.clone();
-
-            std::thread::Builder::new()
-                .name("account-purge".to_string())
-                .spawn(move || loop {
-                    if exit.load(Ordering::Relaxed) {
-                        return;
-                    }
-                    // Simulate purge_slots()
-                    db.purge_slot(slot);
-                    sleep(Duration::from_millis(RACY_SLEEP_MS));
-                })
-                .unwrap()
-        };
-
-        let ancestors: Ancestors = vec![(slot, 0)].into_iter().collect();
-        let t_do_load =
-            start_load_thread(with_retry, ancestors, db, exit.clone(), pubkey, move |_| {
```

Comment:
@ryoqun from what I remember, these tests were essentially testing that if we try to load on a purged note, we panic.

This was because the behavior before was we only removed the storage entry in `purge_slot()` and left the accounts index entry untouched. This meant we would fail to retrieve the account, retry the get from the index, see the same index entry again and panic when we got the same index entry: https://github.com/solana-labs/solana/blob/305d9dd3f4679c8026bdd20b18ec9d84e3b770f0/runtime/src/accounts_db.rs#L2674-L2703

With the current set of changes where the `purge_slot()` call above actually removes the account index entry, which means the `load thread` here will realize the account is deleted and will not retry the account index get, returning `None`. This causes the load thread to panic here: https://github.com/solana-labs/solana/blob/305d9dd3f4679c8026bdd20b18ec9d84e3b770f0/runtime/src/accounts_db.rs#L10184 since the pubkey no longer exists in the index.

Essentially it is no longer possible for the panic to happen because the account index entry is removed before the storage entry is removed, so I have removed these tests.

Reply:
@carllin your reasoning sounds good at first glance. let me think on it more thoughly though. btw, another pretty mind-bending pr from you. :)

Reply:
this your comment is finally making sense... what a brain-teasing pr... lol